Optimizing read_into_buffer helper

usamasaqib opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?
This PR changes the `read_into_buffer` function, used in the https path, to read an http request fragment into memory.
Previously, the function would read a constant size of bytes into memory and zero out a portion of the buffer if an excess amount had been read. This approach was necessitated because `bpf_probe_read` helper function on older kernel versions only allowed a read of constant size.

The change in this PR performs the constant sized read in ebpf, and defers the cleanup operation for the golang consumer.

The PR also replaces the http socket filter with a stub program which just performs a tail call into the actual handler. This is done in order to solve the issue described here: https://datadoghq.atlassian.net/wiki/spaces/NET/pages/2326855913/HTTP#Program-size-limit-for-socket-filters

The number of instructions after this change for two of the effected programs:
| Program | before | after | change |
| ---| --- | --- | --- |
| socket/http_filter | 1873 | 1968 | 95 |
| uprobe/SSL_write | 1852 | 1138 | -714 |

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation
When changing the default clang compiler version for system probe from 11.0.1 to 14.0.5, it was observed that the compiler was generating the bytecode for `read_into_buffer` in a manner which would fail on kernel 4.18.
Two approaches for solving this problem are identified. The first uses code constructs to force the compiler to generate bytecode in a manner acceptable to the 4.18 verifier. This approach is demonstrated in commit [be662b4f0d2bbc3f74f1ec088baca253618e130b](https://github.com/DataDog/datadog-agent/commit/be662b4f0d2bbc3f74f1ec088baca253618e130b).

The alternative approach, presented here, defers the zeroing out for the golang consumer. This allows us to avoid performing the problematic operation in ebpf.
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs
One drawback noted by @p-lambert is that we are exposing junk or potentially sensitive data to code where it should not be exposed..
<!--
* What are the possible side-effects or negative impacts of the code change?
-->
### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

usamasaqib committed:
reading complete size in fragment buffer, and post processing later

usamasaqib committed:
code restructuring

brycekahle commented on pkg/network/http/batch_manager.go:
```diff
@@ -129,5 +129,9 @@ func (m *batchManager) GetPendingTransactions() []httpTX {
 	}
 
+	for _, tx := range transactions {
+		tx.PostProcess()
```
Could we do this in `httpBatch.Transactions()`? That way you cannot forget to call this function and have incomplete/bad data.

brycekahle replied:
I believe that is the only function creates `httpTx` objects, but if there are others we should do the same there too.

brycekahle commented on pkg/network/ebpf/c/http-types.h:
```diff
@@ -52,6 +52,7 @@ typedef struct {
     __u16 response_status_code;
     __u64 response_last_seen;
     char request_fragment[HTTP_BUFFER_SIZE] __attribute__ ((aligned (8)));
+    size_t fragment_sz;
```
`fragment_len` might be a bit easier to understand versus knowing that `sz = size`.

nplanel commented on pkg/network/ebpf/c/https.h:
```diff
@@ -16,7 +16,7 @@ static __always_inline void https_process(conn_tuple_t *t, void *buffer, size_t
     __builtin_memset(&http, 0, sizeof(http));
     __builtin_memcpy(&http.tup, t, sizeof(conn_tuple_t));
-    read_into_buffer((char *)http.request_fragment, buffer, len);
+    http.fragment_sz = read_into_buffer((char *)http.request_fragment, buffer, len);
```
you will need to update http_begin_request() as well to copy fragment_sz to the actual http transaction that will be sent to the userspace later

brycekahle commented on pkg/network/http/model.go:
```diff
@@ -40,6 +40,22 @@ func (k *httpBatchKey) Prepare(n httpNotification) {
+func postProcessFragment(tx *httpTX) {
+	b := *(*[HTTPBufferSize]byte)(unsafe.Pointer(&tx.request_fragment))
+	fSz := tx.fragmentSize()
+	for i := fSz; i < HTTPBufferSize; i++ {
+		b[i] = 0
+	}
```
I wonder if having a global var `zero [HTTPBufferSize]byte` and using `copy(b[fSz:], zero)` would be faster? Maybe worth benchmarking?

usamasaqib committed:
post processing when batch of transactions is accessed

usamasaqib committed:
renaming and cleanup

usamasaqib committed:
updating fragment_len in all places

usamasaqib committed:
fix verifier error

Caluclating read_len earlier in the function and returning it later
causes the compiler to save the value on to the stack, causing the
verifier on older kernels to loose the state information.
http->fragment_len is then set by offseting into the stack and setting
it to read_len.
This causes the verifier to complain about arithmetic between fp and a
register with an unbounded value.
Directly returning the read_len keeps the value in the register, and
preserves the state information.

usamasaqib commented:
> @usamasaqib can you please share with us the number of instructions of the eBPF program before and after the change?

@p-lambert updated the PR description

usamasaqib committed:
reading HTTP_BUFFER_SIZE in 'read_into_buffer_skb' function

usamasaqib commented on pkg/network/ebpf/c/http-buffer.h:
```diff
@@ -3,55 +3,17 @@
+    if (data_size < HTTP_BUFFER_SIZE)
+	    return data_size;
+    else
+	    return HTTP_BUFFER_SIZE-1;
```
If we always return the number of bytes copied, we won't be able to tell the userspace consumer on how much of the buffer is valid data.

HTTP_BUFFER_SIZE-1 is returned so we don't have a buffer overrun in the userspace. This ensures that we will always have a terminating NULL byte in the end of the buffer.

hmahmood replied:
Ok, it is not clear that is why in the code; can you put a comment in please?

Also if the buffer is essentially a NUL terminated string, then is putting a single `NUL` in the ebpf code at the end of the data feasible? That way we could just remove the userspace setting of `NUL`'s?

usamasaqib replied:
I will add the comment.

I agree that we can just put the NULL byte directly in the ebpf code. However, the previous version of read_into_buffer spent significant effort making sure the buffer is zeroed out.
Maybe there is some requirement I am missing?

usamasaqib committed:
revert 'read_into_buffer_skb' change

usamasaqib committed:
fixed read_into_buffer_skb after last revert

usamasaqib committed:
comments added

usamasaqib committed:
attaching a stub program for http socket filter
This is done to bypass memory limitations imposed by the kernel when
attaching a filter to a socket. Using a stub program which branches into
a subprogram allows us to avoid the check which enforces this limit.
See for description of the problem:
https://datadoghq.atlassian.net/wiki/spaces/NET/pages/2326855913/HTTP#Program-size-limit-for-socket-filters
The check enforces that there is enough space in the sockets option
memory buffer:
https://elixir.bootlin.com/linux/v5.4/source/net/core/filter.c#L1198

usamasaqib committed:
HTTP_BUFFER_SIZE increased

usamasaqib committed:
BLOCK_SIZE changed back to original value

brycekahle commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -66,8 +66,17 @@ static __always_inline void read_into_buffer_skb(char *buffer, struct __sk_buff*
+    return HTTP_BUFFER_SIZE <= (skb->len - (u32)offset) ? HTTP_BUFFER_SIZE - 1 : skb->len - (u32)offset;
 }
 
+SEC("socket/http_filter_stub")
```
Can we name this `socket/http_filter_entry`? I think that might clarify the intent.

usamasaqib committed:
revert HTTP_BUFFER_SIZE

nplanel commented on pkg/network/http/model.go:
```diff
@@ -40,6 +40,17 @@ func (k *httpBatchKey) Prepare(n httpNotification) {
 
+func (tx *httpTX) postProcess() {
+	b := *(*[HTTPBufferSize]byte)(unsafe.Pointer(&tx.request_fragment))
+	for i := tx.fragmentSize(); i < HTTPBufferSize; i++ {
```
may a copy of null bytes (with some boundary check before) would be faster `copy(b[tx.fragmentSize():], zeroHttpBuffer[tx.fragmentSize():HTTPBufferSize])`

`var zeroHttpBuffer [HTTPBufferSize]byte`

nplanel commented on pkg/network/http/model.go:
```diff
@@ -40,6 +40,17 @@ func (k *httpBatchKey) Prepare(n httpNotification) {
+}
+
+func (tx *httpTX) fragmentSize() int {
+	return int(tx.fragment_len)
```
could we add telemetry like (h.telemetry.malformed.Inc() / statKeeper:processHTTPPath()) when we see unrealistic fragment size, may later, probably in httpStatKeeper

brycekahle replied:
What is an unrealistic value?

usamasaqib committed:
renamed http socket filter entry point

usamasaqib committed:
Remove passing of fragment_len to userspace

It was noted that it is unnecessary to pass the length of the fragment
buffer to userspace, because the userspace parsing of the HTTP path will
ignore any junk data present in the fragment buffer. This commit removes
the bulk of the changes which were introduced in this PR.

usamasaqib committed:
comments added

usamasaqib committed:
add runtime asset file
